impl(bigquery): Applies SafeGetTo function to common resources

jsrinnn opened the pull request:
1) This PR applies the SafeGetTo() from [PR-12199](https://github.com/googleapis/google-cloud-cpp/pull/12199) to the resources defined in `common_v2_resources.h` that all apis use.

2) it also adds a clang-tidy warning suppression related to recursive fields that are used in the common resources.

FYI: I will be sending around 4 to 5 PRs after this one, that applies the SafeGetTo() 
to the different V2 apis i.e Job, Table, Dataset and Projects.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12208)
<!-- Reviewable:end -->

dbolduc commented on google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc:
```diff
@@ -194,22 +195,36 @@ void from_json(nlohmann::json const& j, Value& v) {
-      case 1:
-        v.value_kind = j.at("valueKind").get<double>();
+      case 1: {
+        double val;
+        SafeGetTo(val, j, "valueKind");
+        v.value_kind = val;
```
you might want a `SafeValueOr` here and below.

https://github.com/googleapis/google-cloud-cpp/pull/12209#discussion_r1273010101

dbolduc commented on google/cloud/bigquery/v2/minimal/internal/json_utils.h:
```diff
@@ -42,12 +42,16 @@ void FromJson(std::chrono::hours& field, nlohmann::json const& j,
 void ToJson(std::chrono::hours const& field, nlohmann::json& j,
             char const* name);
 
+// Suppress recursive clang-tidy warnings
+//
+// NOLINTBEGIN(misc-no-recursion)
```
why are these necessary? `SafeGetTo` does not call itself.

jsrinnn reviewed (commented):
Addressed feedback. Thank You!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12208#-:-NaL7jRJERcSabRpktdK:bh8mug1)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @dbolduc)

___
*[`google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc` line 47 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12208#gh-1273021774:-NaKqIdw4eWchziVWC6q:bs7wumb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/46c61de16f55390357d1870b55f3a9e39af82583/google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc#L47)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

If we call `j.contains(...)` and `SafeGetTo(..., j, ...)`, we are still doing two look ups.

Consider:

```cc
template <typename T>
SafeGetTo(std::shared_ptr<T> value, nlohmann::json const& j, std::string const& key) {
  auto i = j.find(key);
  if (i == j.end()) return;
  if (value == nullptr) value = std::make_shared<T>();
  i->get_to(*value);
}
```

(here and elsewhere)
</blockquote></details>

Done.
___
*[`google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc` line 201 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12208#gh-1273040771:-NaL2L5M8heCj3moGXw-:b-3w4m81) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/46c61de16f55390357d1870b55f3a9e39af82583/google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc#L201)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

you might want a `SafeValueOr` here and below.

https://github.com/googleapis/google-cloud-cpp/pull/12209#discussion_r1273010101
</blockquote></details>

Based on the latest feedback on #12209 (comment), we only want the value to be set if `j.contains()` is true. 
The latest modifications in that PR addresses the above for all the setter member functions.

Since we don't have setters here, I have addressed the above a little differently. 
PTAL and let me know if you think of a better alternative.
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.h` line 47 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12208#gh-1273041388:-NaKvLvj0dBcZ5_n4IlM:bo34b9j) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/46c61de16f55390357d1870b55f3a9e39af82583/google/cloud/bigquery/v2/minimal/internal/json_utils.h#L47)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

why are these necessary? `SafeGetTo` does not call itself.
</blockquote></details>

This gives a clang-tidy error because of their call/usage in `common_v2_resource.cc` with fields that are self-referential.


<!-- Sent from Reviewable.io -->

dbolduc commented on google/cloud/bigquery/v2/minimal/internal/json_utils.h:
```diff
@@ -51,6 +51,28 @@ void SafeGetTo(ResponseType& value, nlohmann::json const& j,
+}
+
+template <typename T>
+bool SafeValueOr(T& value, nlohmann::json const& j, std::string const& key) {
```
We don't need `void SafeGetTo(...)` and `bool SafeValueOr(...)`. We should just modify `SafeGetTo` to return a `bool` like this function does.

https://github.com/googleapis/google-cloud-cpp/blob/916b808c9e918807dd3abd09049dc72e2e1bfd0c/google/cloud/bigquery/v2/minimal/internal/json_utils.h#L48-L53

dbolduc commented on google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc:
```diff
@@ -139,9 +129,8 @@ void from_json(nlohmann::json const& j, StandardSqlDataType& t) {
     auto const index = j.at("sub_type_index").get<int>();
```
(Don't fix in this PR, just an FYI):

This is another double look up. `j.contains(...)` followed by `j.at(...)`.

dbolduc commented on google/cloud/bigquery/v2/minimal/internal/json_utils.h:
```diff
@@ -51,6 +51,28 @@ void SafeGetTo(ResponseType& value, nlohmann::json const& j,
 }
+
+template <typename T>
+std::shared_ptr<T> SafeGetTo(std::shared_ptr<T>& value, nlohmann::json const& j,
```
I think we should return a `bool` to keep the semantics the same.

jsrinnn reviewed (commented):
Addressed all feedback. Thank You!

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12208#-:-NaP5p56-FRcJsjRsk_6:b5jke68)* status: 0 of 3 files reviewed, 6 unresolved discussions (waiting on @dbolduc)

___
*[`google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc` line 129 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12208#gh-1276313183:-NaP-hda7K0XSkOD_N44:b-k513va) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/916b808c9e918807dd3abd09049dc72e2e1bfd0c/google/cloud/bigquery/v2/minimal/internal/common_v2_resources.cc#L129)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

(Don't fix in this PR, just an FYI):

This is another double look up. `j.contains(...)` followed by `j.at(...)`.


</blockquote></details>

Done.

Thanks for catching that. Fixed it here and elsewhere in the file. There should be no more `contains`.
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.h` line 56 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12208#gh-1276325312:-NaOy3ew9S09g4nppxb7:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/916b808c9e918807dd3abd09049dc72e2e1bfd0c/google/cloud/bigquery/v2/minimal/internal/json_utils.h#L56)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

I think we should return a `bool` to keep the semantics the same.
</blockquote></details>

Done.
___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.h` line 68 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12208#gh-1276301412:-NaOy3P02Wefp9sRXT9j:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/916b808c9e918807dd3abd09049dc72e2e1bfd0c/google/cloud/bigquery/v2/minimal/internal/json_utils.h#L68)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

We don't need `void SafeGetTo(...)` and `bool SafeValueOr(...)`. We should just modify `SafeGetTo` to return a `bool` like this function does.

https://github.com/googleapis/google-cloud-cpp/blob/916b808c9e918807dd3abd09049dc72e2e1bfd0c/google/cloud/bigquery/v2/minimal/internal/json_utils.h#L48-L53
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jsrinnn commented:
Friendly Ping.

@dbolduc , can you PTAL and approve when you get a chance if there are no more changes needed? I think I have addressed all the feedback. Thank You!

jsrinnn commented:
Resolved conflicts and rebased from main
